### PR TITLE
Reduce usage of `cardano-api` in `Write.Tx.BalanceSpec`

### PR DESCRIPTION
**Pull request by Anviking**
- Reduce usage of `cardano-api` in `BalanceSpec`

### Issue Number

Follow-up to ADP-2353

### CODE REVIEW

File: lib/wallet/test/unit/Internal/Cardano/Write/Tx/BalanceSpec.hs [hunk 1]
```diff
@@ -1518,52 +1516,39 @@ prop_balanceTransactionValid
...
-            CardanoApi.TxInsCollateral _ [] -> False
-            CardanoApi.TxInsCollateral _ (_:_) -> True
-
-    txFee :: CardanoApi.Tx (CardanoApiEra era) -> CardanoApi.Lovelace
-    txFee (CardanoApi.Tx (CardanoApi.TxBody content) _) =
-        case CardanoApi.txFee content of
-            CardanoApi.TxFeeExplicit _ c -> c
-            CardanoApi.TxFeeImplicit i -> case i of {}
+    hasCollateral :: Tx era -> Bool
+    hasCollateral = not . Set.null
```
Comment (Anviking) [inline lib/wallet/test/unit/Internal/Cardano/Write/Tx/BalanceSpec.hs]: Should just reuse the top level `hasInsCollateral` instead at it has the same type

File: lib/wallet/test/unit/Internal/Cardano/Write/Tx/BalanceSpec.hs [hunk 2]
```diff
@@ -731,11 +733,7 @@ spec_balanceTransaction = describe "balanceTransaction" $ do
                 Left (ErrBalanceTxAssignRedeemers
                         (ErrAssignRedeemersTargetNotFound faultyRedeemer))
   where
-    outputs
-        :: CardanoApi.Tx CardanoApi.BabbageEra
-        -> [Babbage.BabbageTxOut StandardBabbage]
-    outputs (CardanoApi.Tx (CardanoApi.ShelleyTxBody _ body _ _ _ _ ) _) =
-        Write.outputs body
+    outputs = F.toList . view (bodyTxL . outputsTxBodyL)
```
Comment (jonathanknowles) [inline lib/wallet/test/unit/Internal/Cardano/Write/Tx/BalanceSpec.hs:736]: I think having a type signature here might help readability:

```suggestion
    outputs :: IsRecentEra era => Tx era -> [TxOut era]
    outputs = F.toList . view (bodyTxL . outputsTxBodyL)
```

File: lib/wallet/test/unit/Internal/Cardano/Write/Tx/BalanceSpec.hs [hunk 3]
```diff
@@ -862,7 +860,7 @@ balanceTransactionGoldenSpec = describe "balance goldens" $ do
                     dummyTimeTranslation
                     testStdGenSeed
                     ptx
-                combinedUTxO = toCardanoApiUTxO $ mconcat
+                combinedUTxO = mconcat
                     [ view #inputs ptx
                     , fromWalletUTxO walletUTxO
                     ]
```
Comment (jonathanknowles) [inline lib/wallet/test/unit/Internal/Cardano/Write/Tx/BalanceSpec.hs:866]: 👍🏻

File: lib/wallet/test/unit/Internal/Cardano/Write/Tx/BalanceSpec.hs [hunk 4]
```diff
@@ -1349,8 +1350,8 @@ prop_balanceTransactionValid
                         , prop_validSize tx combinedUTxO
                         , prop_minfeeIsCovered tx combinedUTxO
                         , let
-                              minUTxOValue = CardanoApi.Lovelace 999_978
-                              upperBoundCostOfOutput = CardanoApi.Lovelace 1_000
+                              minUTxOValue = Coin 999_978
+                              upperBoundCostOfOutput = Coin 1_000
```
Comment (jonathanknowles) [inline lib/wallet/test/unit/Internal/Cardano/Write/Tx/BalanceSpec.hs:1354]: 👍🏻

File: lib/wallet/test/unit/Internal/Cardano/Write/Tx/BalanceSpec.hs [hunk 5]
```diff
@@ -1437,15 +1438,14 @@ prop_balanceTransactionValid
...
+        :: Coin
+        -> Tx era
         -> UTxO era
         -> Property
     prop_expectFeeExcessSmallerThan lim tx utxo = do
         let fee = txFee tx
         let minfee = minFee tx utxo
-        let unLovelace (CardanoApi.Lovelace x) = x
-        let delta = CardanoApi.Lovelace $ (unLovelace fee) - (unLovelace minfee)
+        let delta = fee <-> minfee
```
Comment (jonathanknowles) [inline lib/wallet/test/unit/Internal/Cardano/Write/Tx/BalanceSpec.hs:1448]: I think we could just use the ordinary group subtraction operator [`~~`](https://hackage.haskell.org/package/groups-0.5.3/docs/Data-Group.html#v:-126--126-) from [`Data.Group`](https://hackage.haskell.org/package/groups-0.5.3/docs/Data-Group.html#t:Group) here:

```suggestion
        let delta = fee ~~ minfee
```

Justification:
- [`MaryValue`](https://input-output-hk.github.io/cardano-ledger/cardano-ledger-mary/Cardano-Ledger-Mary-Value.html#t:MaryValue) and [`Coin`](https://input-output-hk.github.io/cardano-ledger/cardano-ledger-core/Cardano-Ledger-Coin.html#t:Coin) are both instances of [`Group`](https://hackage.haskell.org/package/groups-0.5.3/docs/Data-Group.html#t:Group).
- [`Group`](https://hackage.haskell.org/package/groups-0.5.3/docs/Data-Group.html#t:Group) is a superclass of [`Cardano.Ledger.Val`](https://input-output-hk.github.io/cardano-ledger/cardano-ledger-core/Cardano-Ledger-Val.html#t:Val).
- [`Group`](https://hackage.haskell.org/package/groups-0.5.3/docs/Data-Group.html#t:Group) already provides the functionality we want (group subtraction); IMO we don't gain anything extra by using [`Cardano.Ledger.Val`](https://input-output-hk.github.io/cardano-ledger/cardano-ledger-core/Cardano-Ledger-Val.html#t:Val).

But perhaps there's a particular reason to use `Val` here?

File: lib/wallet/test/unit/Internal/Cardano/Write/Tx/BalanceSpec.hs [hunk 6]
```diff
@@ -1457,14 +1457,13 @@ prop_balanceTransactionValid
...
-        :: CardanoApi.Tx (CardanoApiEra era)
+        :: Tx era
         -> UTxO era
         -> Property
     prop_minfeeIsCovered tx utxo = do
         let fee = txFee tx
         let minfee = minFee tx utxo
-        let unLovelace (CardanoApi.Lovelace x) = x
-        let delta = CardanoApi.Lovelace $ (unLovelace minfee) - (unLovelace fee)
+        let delta = minfee <-> fee
```
Comment (jonathanknowles) [inline lib/wallet/test/unit/Internal/Cardano/Write/Tx/BalanceSpec.hs:1466]: See earlier comment: https://github.com/cardano-foundation/cardano-wallet/pull/4267/files#r1407069258

**Review by jonathanknowles — APPROVED** [review]

File: lib/wallet/test/unit/Internal/Cardano/Write/Tx/BalanceSpec.hs (hunk 5 above)
Comment (Anviking) [reply lib/wallet/test/unit/Internal/Cardano/Write/Tx/BalanceSpec.hs:1448]: As discussed — I'd lean towards using `<->` for the sake of being similar to the ledger implementation. At least for now.

**Pull request merged**